src/python/grpcio_health_checking/grpc_health/v1/_async.py

```diff
@@ -0,0 +1,103 @@
+# Copyright 2020 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Reference implementation for health checking in gRPC Python."""
+
+import asyncio
+import collections
+
+import grpc
+
+from grpc_health.v1 import health_pb2 as _health_pb2
+from grpc_health.v1 import health_pb2_grpc as _health_pb2_grpc
+
+
+class AsyncHealthServicer(_health_pb2_grpc.HealthServicer):
+    """An AsyncIO implementation of health checking servicer."""
+
+    def __init__(self):
+        self._server_status = dict()
+        self._server_watchers = collections.defaultdict(asyncio.Condition)
+        self._gracefully_shutting_down = False
+
+    async def Check(self, request: _health_pb2.HealthCheckRequest, context):
+        status = self._server_status.get(request.service)
+
+        if status is None:
+            await context.abort(grpc.StatusCode.NOT_FOUND)
+        else:
+            return _health_pb2.HealthCheckResponse(status=status)
+
+    async def Watch(self, request: _health_pb2.HealthCheckRequest, context):
+        status = self._server_status.get(request.service)
+
+        if status is None:
+            status = _health_pb2.HealthCheckResponse.SERVICE_UNKNOWN
+
+        try:
+            condition = self._server_watchers[request.service]
+            async with condition:
+                # Responds with current health state
+                await context.write(
+                    _health_pb2.HealthCheckResponse(status=status))
+
+                # Polling on health state changes
+                while True:
+                    await condition.wait()
```

Comment:
The behavior of this implementation differs slightly from the semantics prescribed in the proto:

```
// ... It will then subsequently send a new message whenever
// the service's serving status changes.
```

My reading of that comment is that the client will receive a message each time the health status changes and *only* when the status changes. But depending on scheduling in the event loop, it's possible for the application to very quickly go through the following state changes:

```
SERVING -> NOT_SERVING -> SERVING
```

and the caller of the watch method to observe 
```
... -> SERVING -> SERVING -> ...
```

This is a fundamental shortcoming of concurrent scheduling. Very similar to the quirks of the channel state API we discussed not too long ago. I think it's okay to swallow states that only last for negligible periods of time. So we should filter out duplicate states to account for that.

Not sure how the synchronous health servicer works in this respect since `threading.Condition` should be susceptible to the same race, but it might be worth looking into.